### PR TITLE
New recordings CSV export API endpoint 

### PR DESCRIPTION
**Pull request by mjs**
This new endpoint which takes the same query parameters as the main query API but generates a CSV. This CSV includes extra information such as details of the most recently played audio bait event.

To faciltate this change the recording query logic has been factored out into its own query builder object. The creation of the tests for this feature lead to various improvements in the API server test infrastructure.

### CODE REVIEW

File: api/V1/Recording.js [hunk 1]
```diff
@@ -181,6 +184,31 @@ module.exports = (app, baseUrl) => {
...
+   * @apiName Report
+   * @apiGroup Recordings
+   * @apiDescription Parameters are as per GET /api/V1/recordings. On
+   * success (status 200), the response body will contain CSV
+   * formatted details of the selected recordings.
+   *
+   * @apiUse V1UserAuthorizationHeader
+   * @apiUse QueryParams
+   * @apiParam {JSON} [tags] As per GET /api/V1/recordings
+   * @apiParam {String} [tagMode] As per GET /api/V1/recordings
```
Comment (CameronRP) [inline api/V1/Recording.js]: You can use `apiDefine` here.

Comment (mjs) [reply api/V1/Recording.js]: I'll move tags and tagMode into QueryParams

File: api/V1/recordingUtil.js [hunk 2]
```diff
@@ -46,30 +49,179 @@ function makeUploadHandler(mungeData) {
...
+    request.user,
+    request.filterOptions
+  );
+
+  // Our DB schema doesn't allow us to easily get from a audio event
+  // recording to a audio file name so do some work first to look these up.
+  const audioEvents = new Map();
+  const audioFileIds = new Set();
+  for (const r of result) {
+    const event = findLatestEvent(r.Device.Events);
```
[the remark this answers is not in the document]
Comment (mjs) [reply api/V1/recordingUtil.js]: The where clause of the include for audio events only selects audio events that occured within 30mins of the recording's timestamp. This function is just finding the latest out of the ones that get returned (typically 1 or 2). 

They don't seem to come back in a predictable order and tried to get Sequelise to order them but that did work - hence this function.

Your comment has made me think of a gap in the tests though.

File: api/V1/recordingUtil.js [hunk 3]
```diff
@@ -46,30 +49,179 @@ function makeUploadHandler(mungeData) {
...
+
+  // Bulk look up file details of played audio events.
+  const audioFileNames = new Map();
+  for (const f of await models.File.getMultiple(Array.from(audioFileIds))) {
+    audioFileNames[f.id] = f.details.name;
+  }
+
+  console.log(audioEvents);
+  console.log(audioFileIds);
+  console.log(audioFileNames);
```
[the remark this answers is not in the document]
Comment (mjs) [reply api/V1/recordingUtil.js]: Good catch. Thanks.

**Pull request merged**